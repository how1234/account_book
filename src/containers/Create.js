import React from 'react'
import CategorySelector from '../components/CategorySelector'
import {Tabs,Tab} from '../components/Tab'
import {testCategories} from '../testData'
import {TYPE_INCOME,TYPE_OUTCOME} from '../ultilities'
import PriceForm from '../components/PriceForm'
import WithContext from '../WithContext'
import {withRouter} from 'react-router-dom'
import {AppContext} from '../App'

const tabsText = [TYPE_OUTCOME,TYPE_INCOME]

class Create extends React.Component {
    constructor(props){
        super(props)
        const {id} = props.match.params
        const {categories,items} = props.data
        this.state = {
            selectedTab : (id && items[id]) ? categories[items[id].cid].type: TYPE_OUTCOME,
            selectedCategory:(id && items[id])? categories[items[id].cid]  : null,
            validation:true
        }
    }
    formSubmit = (data,isEditMode) => {

        if(!this.state.selectedCategory){
            this.setState({
                validation:false
            })
            return
        }
        if(!isEditMode) {
            this.props.actions.createItem(data,this.state.selectedCategory.id)
        }else{
            this.props.actions.updateItem(data,this.state.selectedCategory.id)
        }
        this.props.history.push('/')
    }
    cancelSubmit = () => {
        this.props.history.push('/')
    }
    tabChange = (index) => {
        this.setState({
            selectedTab: tabsText[index]
        })
    }
    selectCategory = (category) =>{
        this.setState({
            selectedCategory:category
        })
    }   
    render(){
        const {data} = this.props
        const {items,categories} = data
        const {id} = this.props.match.params
        const editItem = (id && items[id]) ? items[id] : {}
        const {selectedTab,validation} = this.state
        const filteredCategories = Object.keys(categories).filter(id=> categories[id].type === selectedTab)
        .map(id =>categories[id]) 
        const tabIndex = tabsText.findIndex(text => text === selectedTab)
        return(
            
            <div className="create-page py-3 px3 rounded mt-3" style={{background:'#fff'}}>
                <Tabs activeIndex={tabIndex} onTabChange={this.tabChange} >
                    <Tab>Income</Tab>
                    <Tab>Outcome</Tab>
                </Tabs>
                <CategorySelector categories={filteredCategories} 
                onSelectCategory={this.selectCategory}
                selectedCategory={this.state.selectedCategory} />
                <PriceForm 
                    onFormSubmit = {this.formSubmit}
                    onCancelSubmit = {this.cancelSubmit}
                    item={editItem}
                />
                {!validation && <div className="alert alert-danger mt-5" role="alert">
                        Please choose a valid category
                    </div>}
            </div>
                   
        )
    }
}

export default withRouter(WithContext(Create))