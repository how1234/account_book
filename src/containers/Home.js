import React from 'react'
import logo from '../logo.svg'
import Ionicon from 'react-ionicons'

import {withRouter} from 'react-router-dom'



import PriceList from '../components/PriceList'

import TotalPrice from '../components/TotalPrice'
import {LIST_VIEW,CHART_VIEW,TYPE_INCOME,TYPE_OUTCOME,parse2YearAndMonth,padMonth} from '../ultilities'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn';
import {Tabs,Tab} from '../components/Tab'
import {testCategories,testItems} from '../testData'
import WithContext from '../WithContext'
import Loader from '../components/Loader'


    const tabsList = [
        LIST_VIEW,
        CHART_VIEW
    ]
  class Home extends React.Component{
      constructor(props){
          super(props)
          this.state={
            tabView:tabsList[0]
          }
      }
      componentDidMount() {
        this.props.actions.getInitialData()
      }
      
      changeView = (index) =>{
        this.setState({
            tabView:tabsList[index]
        })
      }
      changeDate = (year,month) =>{
          this.props.actions.selectNewMonth(year,month)
      }
      editItem = (modifiedItem) =>{
        this.props.history.push(`/edit/${modifiedItem.id}`)
      }
      createItem = () => {
        this.props.history.push('/create')
        
      }
      deleteItem = (deletedItem) => {
          console.log(this.props.actions.deleteItem(deletedItem))
        this.props.actions.deleteItem(deletedItem)
      }
      render() {
          const {data} = this.props
          const {items,categories,currentDate,isLoading} = data
          const {tabView} = this.state
        
       
          const itemsWithCategory = Object.keys(items).map(id=>{
              items[id].category = categories[items[id].cid]
              return items[id]
          })

          let totalIncome = 0,totalOutcome = 0;

          itemsWithCategory.forEach( (item) => {
              if(item.category.type === TYPE_OUTCOME){
                    totalOutcome += item.price 
              }else{
                    totalIncome += item.price
              }
          })


          
          return(
           
                <React.Fragment>
                    <header className="App-header pb-5">
                        <div className="row mb-5 justify-content-center">
                            <img src={logo} className="App-logo" alt="logo" />
                        </div>
                            <div className="row">
                                <div className="col">
                                    <MonthPicker 
                                        year={currentDate.year} 
                                        month={currentDate.month} 
                                        onChange={this.changeDate}/>
                                </div>

                                <div className="col">
                                    <TotalPrice 
                                    income={totalIncome}
                                    outcome={totalOutcome}/>

                        
                                </div>

                            </div>
                        
                        </header>

                        <div className="container-area py-3 px-3">
                        {isLoading && <Loader/>}
                        {!isLoading && 
                          <React.Fragment>
                        <Tabs activeIndex={0} onTabChange={this.changeView}>
                                <Tab>
                                    <Ionicon 
                                    className="rounded-circle mr-2" 
                                    fontSize="25px"
                                    color={'#007bff'}
                                    icon='ios-paper'
                                    />
                                    List
                                </Tab>
                                <Tab>
                                    <Ionicon 
                                    className="rounded-circle mr-2" 
                                    fontSize="25px"
                                    color={'#007bff'}
                                    icon='ios-pie'
                                    />
                                    Chart
                                </Tab>
                        </Tabs>
                            <CreateBtn onClick={this.createItem}></CreateBtn>    

                            {tabView ===LIST_VIEW && 
                            <PriceList items={itemsWithCategory}
                            onEditItem={this.editItem}
                            onDeleteItem={this.deleteItem}
                            ></PriceList>} 

                            {tabView ===CHART_VIEW && 
                            <h1 className="chart-title">Here is for chart</h1>} 
                            </React.Fragment>
                        }
                        </div>
                        
                        <div>
                        </div>

                        
                        
                </React.Fragment>
               
          )
      }
  }

  export default withRouter(WithContext(Home))