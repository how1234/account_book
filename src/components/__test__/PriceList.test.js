import React from 'react'
import { shallow } from 'enzyme'
import PriceList from '../PriceList'
import {items,categories}  from '../../containers/Home'
import Ionicon from 'react-ionicons'

const itemsWithCategory = items.map(item=>{
    item.category = categories[item.cid]
    return item
})

const props = {
    items:itemsWithCategory,
    onEditItem: jest.fn(),
    onDeleteItem: jest.fn()
}

let wrapper
describe('test PriceList component', ()=>{
    beforeEach( () => {
        wrapper = shallow(<PriceList {...props}/>)

    })

    it('should render the component to macth snapshow', ()=>{
        expect(wrapper).toMatchSnapshot()
    })

    it('should render correct price length',()=>{
        expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
    })
    it('should render correct icon and price for each list',()=>{
        const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
        expect(iconList.length).toEqual(3)
        expect(iconList.first().props().icon).toEqual(itemsWithCategory[0].category.iconName)
    })
    
    it('should trigger the correct callback functions', ()=>{
        const firstItem = wrapper.find('.list-group-item').first()
        firstItem.find('a').first().simulate('click')
        expect(props.onEditItem).toHaveBeenCalledWith(itemsWithCategory[0])

        firstItem.find('a').last().simulate('click')
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0])
    })
})