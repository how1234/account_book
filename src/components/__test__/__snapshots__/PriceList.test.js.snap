// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to macth snapshow 1`] = `ShallowWrapper {}`;
